### DISCUSSION TITLE
[GBA] Anyone able to dump the save of FLASH1M repros

### DISCUSSION THREAD

**Discussion by KobeOne2** [sanni/cartreader#1019] [Need help]
I attempted to obtain the save from a repro of Pokemon Emerald, but it appears that the save is only valid for the first half. The second half of the dumped save seems to be a replica of the first. So, may there be a problem with bank swapping? The cart reader works fine as I have no issue when dumping the save from an official version of Emerald.

The cartridge in question resembles the following:
![cartridge](https://github.com/user-attachments/assets/9972f984-b867-473d-84d1-3496b368a7d1)

**Comment by sanni** [comment]
In case you want to tinker with it yourself, FlashGBX does it like so:
````
if save_size > bank_size:
					if args["save_type"] == 8 or agb_flash_chip in (0xBF5B, 0xFFFF): # Bootleg 1M
						dprint("Switching to bootleg save bank {:d}".format(bank))
						self._cart_write(0x1000000, bank)
					elif args["save_type"] == 5: # FLASH 1M
						dprint("Switching to FLASH bank {:d}".format(bank))
						cmds = [
							[ 0x5555, 0xAA ],
							[ 0x2AAA, 0x55 ],
							[ 0x5555, 0xB0 ],
							[ 0, bank ]
						]
						self._cart_write_flash(cmds)
					else:
						dprint("Unknown bank switching method")
					time.sleep(0.05)
````
https://github.com/lesserkuma/FlashGBX/blob/master/FlashGBX/LK_Device.py#L3119

Here is where the OSCR switches banks: 
````
case 5:
          // 1M FLASH (divided into two banks)
          switchBank_GBA(0x0);
          setROM_GBA();
          readFLASH_GBA(1, 65536, 0);
          switchBank_GBA(0x1);
          setROM_GBA();
          readFLASH_GBA(0, 65536, 65536);
          break;
````
https://github.com/sanni/cartreader/blob/master/Cart_Reader/GBA.ino#L112

So you could try replacing:
````
switchBank_GBA(0x1); 
````
with:
````
writeWord_GBA(0x1000000, 0x1);
````

Maybe it works, maybe you will have to dig a little deeper into the source code of FlashGBX to understand how to switch the bank of the 1M repro flash save type.
